modules/infra/api/gr_infra.h

```diff
@@ -33,25 +33,39 @@
 #define GR_IFACE_SET_FLAGS GR_BIT64(0)
 #define GR_IFACE_SET_MTU GR_BIT64(1)
 #define GR_IFACE_SET_NAME GR_BIT64(2)
-#define GR_IFACE_SET_VRF GR_BIT64(3)
+#define GR_IFACE_SET_MODE GR_BIT64(3)
+#define GR_IFACE_SET_VRF GR_BIT64(4)
+#define GR_IFACE_SET_DOMAIN GR_BIT64(5)
 
 #define GR_IFACE_ID_UNDEF 0
 
 #define GR_VRF_ID_ALL UINT16_MAX
 
+typedef enum gr_iface_mode {
+	GR_IFACE_MODE_L3 = 0,
+	GR_IFACE_MODE_XCONNECT,
+	GR_IFACE_MODE_COUNT
+} __attribute__((mode(QI))) gr_iface_mode_t;
+
 // Generic struct for all network interfaces.
 struct gr_iface {
 	uint16_t id; // Interface unique index.
 	uint16_t type; // Interface type. Uses values from GR_IFACE_TYPE_*.
 	uint16_t flags; // Interface flags. Bit mask of GR_IFACE_F_*.
 	uint16_t state; // Interface state. Bit mask of GR_IFACE_S_*.
 	uint16_t mtu; // Maximum transmission unit size (incl. headers).
-	uint16_t vrf_id; // L3 addressing and routing domain
+	gr_iface_mode_t mode;
+	union {
+		uint16_t vrf_id; // L3 addressing and routing domain
+		uint16_t domain_id; // L2 xconnect peer interface id
+	};
 #define GR_IFACE_NAME_SIZE 64
 	char name[GR_IFACE_NAME_SIZE]; // Interface name (utf-8 encoded, nul terminated).
 	uint8_t info[256]; // Type specific interface info.
 };
 
+void register_interface_mode(gr_iface_mode_t mode, const char *next_node);
```

Comment:
This should go in the datapath headers. Not to be exposed in the API for clients.